File: java/src/main/java/it/giannibombelli/wsc2026/payment/domain/events/PaymentResultEvents.java
package it.giannibombelli.wsc2026.payment.domain.events;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.common.domain.primitive.ClientReference;
import it.giannibombelli.wsc2026.common.domain.primitive.Description;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.payment.domain.payment.PaymentId;

public sealed interface PaymentResultEvents extends PaymentEvent
    permits PaymentResultEvents.PaymentAccepted, PaymentResultEvents.PaymentRejected, PaymentResultEvents.PaymentExpired {

    record PaymentAccepted(
        PaymentId aggregateId,
        ClientReference clientReference,
        Money amount
    ) implements PaymentResultEvents {
        public PaymentAccepted {
            Require.requireArgument(aggregateId, "paymentId");
            Require.requireArgument(clientReference, "clientReference");
            Require.requireArgument(amount, "amount");
        }
    }

    record PaymentRejected(PaymentId aggregateId, ClientReference clientReference, Money amount,
                           Description reason) implements PaymentResultEvents {
        public PaymentRejected {
            Require.requireArgument(aggregateId, "paymentId");
            Require.requireArgument(clientReference, "clientReference");
            Require.requireArgument(amount, "amount");
            Require.requireArgument(reason, "reason");
        }
    }

    record PaymentExpired(PaymentId aggregateId, ClientReference clientReference, Money amount) implements PaymentResultEvents {
        public PaymentExpired {
            Require.requireArgument(aggregateId, "paymentId");
            Require.requireArgument(clientReference, "clientReference");
            Require.requireArgument(amount, "amount");
        }
    }
}

File: typescript/src/payment/domain/events/paymentResultEvents.ts
// Eventi di risultato del ciclo di vita di un pagamento.

import type { Event } from '@/common/domain/model/event.js';
import type { ClientReference } from '@/common/domain/primitive/clientReference.js';
import type { Description } from '@/common/domain/primitive/description.js';
import type { Money } from '@/common/domain/primitive/money.js';
import type { PaymentId } from '../payment/paymentId.js';
import { requireArgument } from '@/common/utils/requireArgument.js';

export type PaymentAccepted = Event<'PaymentAccepted', PaymentId> & {
  readonly clientReference: ClientReference;
  readonly amount: Money;
};

export function paymentAccepted(
  aggregateId: PaymentId,
  clientReference: ClientReference,
  amount: Money
): PaymentAccepted {
  requireArgument(aggregateId, 'paymentId');
  requireArgument(clientReference, 'clientReference');
  requireArgument(amount, 'amount');
  return { kind: 'PaymentAccepted', aggregateId, clientReference, amount };
}

export type PaymentRejected = Event<'PaymentRejected', PaymentId> & {
  readonly clientReference: ClientReference;
  readonly amount: Money;
  readonly reason: Description;
};

export function paymentRejected(
  aggregateId: PaymentId,
  clientReference: ClientReference,
  amount: Money,
  reason: Description
): PaymentRejected {
  requireArgument(aggregateId, 'paymentId');
  requireArgument(clientReference, 'clientReference');
  requireArgument(amount, 'amount');
  requireArgument(reason, 'reason');
  return { kind: 'PaymentRejected', aggregateId, clientReference, amount, reason };
}

export type PaymentExpired = Event<'PaymentExpired', PaymentId> & {
  readonly clientReference: ClientReference;
  readonly amount: Money;
};

export function paymentExpired(
  aggregateId: PaymentId,
  clientReference: ClientReference,
  amount: Money
): PaymentExpired {
  requireArgument(aggregateId, 'paymentId');
  requireArgument(clientReference, 'clientReference');
  requireArgument(amount, 'amount');
  return { kind: 'PaymentExpired', aggregateId, clientReference, amount };
}

export type PaymentResultEvent = PaymentAccepted | PaymentRejected | PaymentExpired;
